dpm/client/__init__.py

```diff
@@ -49,9 +77,67 @@ def validate(self):
         # should we really check this here? Good question i think ...
         for idx, resource in enumerate(self.datapackage.resources):
             if not exists(resource.local_data_path):
-                raise DpmException('Resource at index %s with name % and path %s does not exist on disk' % (
+                raise DpmException('Resource at index %s with name %s and path %s does not exist on disk' % (
                     idx, resource.name, resource.local_data_path)
                     )
 
         return True
 
+    def ensure_auth(self):
+        """
+        Get auth token from the server using credentials. Token can be used in future
+        requests to the server.
+
+        # TODO: is this private or useful outside?
+        # TODO: refresh expired token
+        """
+        if self.token:
+            return self.token
+
+        if not (self.username and self.password):
+            raise MissingCredentialsError('User credentials are missing.')
+
+        #echo('Authenticating ... ', nl=False)  # TODO: logging
+        authresponse = self._apirequest(
+                method='POST',
+                url='%s/api/auth/token' % self.server,
+                json={'username': self.username, 'secret': self.password})
+
+        self.token = authresponse.json().get('token')
+        if not self.token:
+            raise AuthError('Server did not return auth token')
+
+        return self.token
+
+    def _apirequest(self, method, *args, **kwargs):
+        """
+        General request-response processing routine for dpr-api server.
+
+        :return:
+            Response -- requests.Response instance
+
+        """
+        methods = {
+            'POST': requests.post,
+            'PUT': requests.put,
+            'GET': requests.get,
+            'DELETE': requests.delete
+        }
+
+        headers = kwargs.pop('headers', {})
+        if self.token:
+            headers.setdefault('Authorization', 'Bearer %s' % self.token)
+
+        response = methods.get(method)(*args, headers=headers, **kwargs)
+
+        try:
```

Comment:
Because the python2 throws vary generic ValueError on the malformed json. Caller of this method will have hard time to handle exception properly as ValueError could be raised from somewhere else, like encoding what he gave us as parameters.